### PR TITLE
Return all symbols from a kallsym

### PR DESCRIPTION
**Pull request by ThinkerYzu1**
Fix issue #15

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -448,14 +448,19 @@ impl SymResolver for ElfResolver {
...
 
-    fn find_symbol(&self, addr: u64) -> Option<(&str, u64)> {
+    fn find_symbols(&self, addr: u64) -> Vec<(&str, u64)> {
         let off = addr - self.loaded_address + self.loaded_to_virt;
-        let parser = self.get_parser()?;
+        let parser = self.get_parser();
+        if parser.is_none() {
+            return vec![];
+        }
+        let parser = parser.unwrap();
```
Comment (danielocfb) [inline src/lib.rs:420]: Nit: Use `if let` instead of unwrapping?
```rust
let parser = if let Some(parser) = self.get_parser() {
  parser
} else {
  return Vec::new()
};
```

File: src/lib.rs [hunk 2]
```diff
@@ -1139,41 +1144,49 @@ impl BlazeSymbolizer {
...
-                    vec![SymbolizedResult {
-                        symbol: String::from(sym),
-                        start_address: start,
-                        path: linfo.path,
-                        line_no: linfo.line_no,
-                        column: linfo.column,
-                    }]
+                if res_syms.is_empty() {
+                    if linfo.is_some() {
+                        let linfo = linfo.unwrap();
```
Comment (danielocfb) [inline src/lib.rs:1118]: Same suggestion as above.

File: src/lib.rs [hunk 3]
```diff
@@ -1139,41 +1144,49 @@ impl BlazeSymbolizer {
...
+                        vec![SymbolizedResult {
+                            symbol: "".to_string(),
+                            start_address: 0,
```
Comment (danielocfb) [inline src/lib.rs:1121]: I don't think we should use any sentinel values like this. If there is the potential for data not being present then we should use `Option` instead to make this explicit, in my opinion. Behavior like this is precisely where problems such as https://github.com/libbpf/blazesym/issues/15 are rooted in.

File: src/lib.rs [hunk 4]
```diff
@@ -1139,41 +1144,49 @@ impl BlazeSymbolizer {
...
-                    let (sym, start) = sym.unwrap();
-                    vec![SymbolizedResult {
-                        symbol: String::from(sym),
-                        start_address: start,
-                        path: "".to_string(),
-                        line_no: 0,
-                        column: 0,
-                    }]
+                    let mut results = vec![];
+                    for sym in res_syms {
```
Comment (danielocfb) [inline src/lib.rs:1131]: Not really sure I understand the signature of this `symbolize` method. Why does it return nested vectors and not just a single one?

File: src/ksym.rs [hunk 5]
```diff
@@ -91,29 +91,69 @@ impl KSymResolver {
         }
     }
 
-    pub fn find_address_ksym(&self, addr: u64) -> Option<&Ksym> {
+    pub fn find_address_ksym(&self, addr: u64) -> Vec<&Ksym> {
```
Comment (danielocfb) [inline src/ksym.rs:94]: Why are we using plural in one of the names (i.e., renaming it) but not here? Seems inconsistent, given that both methods can now return multiple addresses?

File: src/ksym.rs [hunk 6]
```diff
@@ -91,29 +91,69 @@ impl KSymResolver {
...
             }
         }
+        assert!(
+            l == 0 || self.syms[l].addr <= addr,
+            "left addr <= addr < right addr"
+        );
+        assert!(
+            r == self.syms.len() || addr < self.syms[r].addr,
+            "left addr <= addr < right addr"
+        );
```
Comment (danielocfb) [inline src/ksym.rs:123]: These are effectively assertions making claims about user provided data, correct? I don't think we should panic on potentially malformed input, but handle it gracefully, no?

File: src/ksym.rs (hunk 5 above)
Comment (danielocfb) [inline src/ksym.rs:94]: Would it be possible that we use https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search_by instead of rolling our own binary search here, twice?

I don't really understand what `dropped_l`'s purpose is and what high level problem we solve with those two loops, to be honest. Can you explain?

**Review by danielocfb — CHANGES_REQUESTED** [review]

File: src/ksym.rs [hunk 7]
```diff
@@ -91,29 +91,69 @@ impl KSymResolver {
...
+        // Seek `l` to the first symbol with the same addr as the
+        // symbol at the current position given by l.
+        let start_addr = self.syms[l].addr;
+        if self.syms[dropped_l].addr == start_addr {
+            assert_eq!(dropped_l, 0);
+            l = dropped_l;
+        } else {
+            // Do binary searching in the range (dropped_l, l] to find the
+            // first symbol with the same address as lth symbol.
+            while dropped_l < (l - 1) {
```
Comment (anakryiko) [inline src/ksym.rs:134]: so this function can be logically split into two parts:
1) find rightmost ksym with ksym.addr <= addr
2) starting from that symbol, append all symbols that have ksym.addr <= addr and stop at the first ksym.addr > addr

Is that right? This should be doable with one simple and elegant binary search, something like this (in C and sort of pseudo-code, but hopefully will give an idea, but note that I haven't even tried to compile this):

```
int n = self.syms.len();
int l = 0, r = n - 1;
while (l < r) {
  int m = l + (r - l + 1) / 2;
  if (self.syms[m].addr <= addr)
    l = m;
  else
    r = m - 1;
}
for (; l < n && self.syms[l].addr <= addr; l++) {
    results.push(&self.syms[l]);
}
```

Note that you don't need any extra pre- or post-checks, this should handle all the cases cleanly.

Few points here that are missing from current implementation:

a) the invariant is that index `l` always points to ksym with ksym.addr <= addr, and binary search makes sure we always try to pull this left index to the right, if possible. One corner case is when leftmost ksym already has too big addr. We can specially check that condition before the loop, but in this case it is not necessary because the loop that add elements tests this condition anyway. So binary search will end with l == r == 0 and following loop will add no elements.
b) `while (l < r)` makes sure we don't miss any elements, we'll stop only when l == r, you code has `l < r - 1` which while might be correct if you take into account all the other checks, doesn't inspire confidence at all
c) `int m = l + (r - l + 1) / 2;` makes sure that if we have l == r-1, we always try right element and if it satisfies condition we update to point to that right element
d)  `r = m - 1;` in combination with above means we always make progress

There are literally 5 meaningful lines in any binary search implementation and each line is extremely important. But also slight variations of those 5 lines define different variations of what you need to find (leftmost vs rightmost element with some property, etc)

Comment (ThinkerYzu1) [reply src/ksym.rs:134]: After the while-loop, all symbols with an address that is less or equal to `addr` will be placed before `l` or at `l`, and all address bigger than `addr` will at the right side of `l`.  The address of the corresponding symbols of `addr` is always less or equal to `addr`. That means the corresponding symbols will be at the left side of `l`. `l` will point to the "last" one of the symbols with the same address.

Does that make sense?

Comment (ThinkerYzu1) [reply src/ksym.rs:134]: After the while-loop, all symbols with addresses that are less than or equal to `addr` will be after the left size of `l` except the last one.  All symbols with addresses that are bigger than `addr` will be at the right size of `l`.  And, the address of the corresponding symbols of `l` is always less or equal to `addr`.  That mean `l` points to the last one, the leftmost, of the corresponding symbols.

Does that make sense?

We can push symbols into the vector in the reversed order instead.  Then, we don't need to seek `l` to the first matched symbol.

File: src/lib.rs (hunk 4 above)
Comment (ThinkerYzu1) [reply src/lib.rs:1131]: It symbolizes a list of addresses.  So, it needs an entry for each address.  However, several symbols can be at the same address for reasons.  One of the reasons is optimization, LTO for example, can merge functions.

File: src/ksym.rs (hunk 5 above)
Comment (ThinkerYzu1) [reply src/ksym.rs:94]: agree!

Comment (ThinkerYzu1) [reply src/ksym.rs:94]: According to the document,  it says
```
If the value is found then [Result::Ok](https://doc.rust-lang.org/std/result/enum.Result.html#variant.Ok) is returned, 
containing the index of the matching element. If there are multiple matches, then any one of the matches could be 
returned.
```

The problem here is the potential of having multiple symbols at the same address.

File: src/ksym.rs (hunk 6 above)
Comment (ThinkerYzu1) [reply src/ksym.rs:123]: The assertions are making claims of the behavior of `KSymResolver`.  The symbols are sorted by `KSymResolver`.  These two assertions claim the facts that the later block depends on.

File: src/ksym.rs (hunk 5 above)
Comment (danielocfb) [reply src/ksym.rs:94]: What I am saying is that it could be used to find the starting position, no? I.e., the one where 
> all symbols with addresses that are less than or equal to addr will be at the left side of `l` except the last one. All symbols with addresses that are bigger than addr will be at the right side of `l`

File: src/ksym.rs (hunk 7 above)
Comment (anakryiko) [reply src/ksym.rs:134]: Not exactly clear which part of my comment you are replying to.

But looking at the original code again, I think I got the problem slightly wrong assuming we want to find all the symbols that might cover provided addr. Now I think we only want to find exact matches, right? So it requires only a slight change to my example:

```
int n = self.syms.len();
int l = 0, r = n - 1;
while (l < r) {
  int m = l + (r - l) / 2;
  if (self.syms[m].addr < addr)
    l = m + 1;
  else
    r = m;
}
for (; l < n && self.syms[l].addr == addr; l++) {
    results.push(&self.syms[l]);
}
```

I updated few details:
a) `int m = l + (r - l) / 2;` so we try left element once we get to two-element range
b)
```
  if (self.syms[m].addr < addr)
    l = m + 1;
  else
    r = m;
```
Here invariant is that `r` contains leftmost symbol with `ksym.addr >= addr`

c) `for (; l < n && self.syms[l].addr == addr; l++) {` so that we only add ksyms with exact addr match, if there is any.

As I said, there are few parts to each binary search implementation which define exact algorithm details, but the overall structure is always the same.

Please let me know if you think this solves the problem.

My big point here is that nested binary searches are not good and elegant solution, so we should avoid it if we can.

File: src/ksym.rs (hunk 5 above)
Comment (ThinkerYzu1) [reply src/ksym.rs:94]: If you are saying to replace the first while-loop with the built-in binary search, yes.  However, we will miss dropped_l.  That means we need to search in the range (0..l] instead of the range (dropped_l..l].

File: src/ksym.rs (hunk 7 above)
Comment (ThinkerYzu1) [reply src/ksym.rs:134]: kallsyms contains (start address, symbol name) pairs. The function is to find a symbol A followed by symbol B where the address of A <= target_address < the address of B, and includes all symbols with the same start address as A.  So, we are not doing exactly matching.

Comment (ThinkerYzu1) [reply src/ksym.rs:134]: The optimization here seems to be very confusing to people.  Should we go back to the most simple way by doing a backward linear search?

File: src/ksym.rs (hunk 5 above)
Comment (danielocfb) [reply src/ksym.rs:94]: The methods accepts a `FnMut`, meaning you should be able to write to the environment from it. I.e., you can close over the `dropped_l` variable and modify it from the closure body, while abstracting away the finer binary search details.

File: src/ksym.rs (hunk 7 above)
Comment (anakryiko) [reply src/ksym.rs:134]: > kallsyms contains (start address, symbol name) pairs, and we sort them. The function is to find a symbol A followed by symbol B where the address of A <= target_address < the address of B, and includes all symbols with the same start address as A. So, we are not doing exactly matching.

ok, then my first example should do the job and replace this entire function. The only extra thing you'd need is in the for() loop that goes after while() loop, in addition to checking that ksym.addr <= addr, we need to check that next ksym.addr starts after addr (because kallsyms doesn't store the size of the symbol).

WDYT?

Comment (anakryiko) [reply src/ksym.rs:134]: @ThinkerYzu1 ok, after re-reading your last reply and the one quoted below I now understand what you are trying to say. If I actually understand :) then we need to change for() loop after while() loop in my first example to go backwards as long as `ksym[l].addr <= addr && addr < ksym[l + 1].addr`. Is that right? So what you said, just add elements in backwards order.

> After the while-loop, all symbols with addresses that are less than or equal to `addr` will be at the left side of `l` except the last one. All symbols with addresses that are bigger than `addr` will be at the right side of `l`. And, the address of the corresponding symbols of `l` is always less or equal to `addr`. That mean `l` points to the last one, the leftmost, of the corresponding symbols.
> 
> Does that make sense?
> 
> We can push symbols into the vector in the reversed order instead. Then, we don't need to seek `l` to the first matched symbol.

yep.

My **main point** I was trying to make in this comment thread is that this is the problem that should be solvable with one clean binary search loop. Nested loop and all the extra checks should be avoided if possible.

WDYT?

File: src/ksym.rs (hunk 5 above)
Comment (danielocfb) [reply src/ksym.rs:94]: But...now that it has become clearer what the code is trying to do, the method does not support looking at more than one element at a time, as in Andrii's condition: `ksym[l].addr <= addr && addr < ksym[l + 1].addr`. So if that's required, we have to hand roll. If we wanted to use it, we'd have to scan both directions for additional matches. In that case it's more of a trade-off. I am fine either way.

File: src/ksym.rs (hunk 7 above)
Comment (ThinkerYzu1) [reply src/ksym.rs:134]: Agree! Let's add symbols in backward order after the first loop.

File: src/lib.rs (hunk 3 above)
Comment (ThinkerYzu1) [reply src/lib.rs:1121]: Not really.  #15 is about multiple symbols with the same address, but we returned only one of them.
However, I agree that the `Option` is better here.

File: src/ksym.rs [hunk 8]
```diff
@@ -91,29 +91,26 @@ impl KSymResolver {
...
-                r = v;
+                r = m;
             } else {
-                l = v;
+                l = m + 1;
             }
         }
-
-        Some(&self.syms[l])
+        // `self.syms[l - 1].addr <= addr < self.syms[l].addr`.
```
Comment (ThinkerYzu1) [inline src/ksym.rs:108]: The new implementation is basically the first part of the previous revision, but remove the check of the exact match.  It is almost the same as what @Andrii provided; however, I changed the condition of moving left and right a little bit.

Here it moves left (`l`) when `addr >= sym.addr` instead of `addr > sym.addr`.
Moving left when `addr > sym.addr` would make 
 - `l` points at **the first one of the matched symbols** for the **exactly matched** cases, while 
 - `l` points to **the position after the matched symbols** for the cases that the address of the matched symbols is **smaller**.

**Review by danielocfb — COMMENTED** [review]
> Previously, it returned only one of the symbols of the given address.
> Returning all symbols fixes the test case ksym_resolver_load_find.

The test may be passing now, but it's passing despite arguably doing the wrong thing. Because as I said earlier and laid out in https://github.com/libbpf/blazesym/issues/15, we are still treating 0 addresses wrongly. So what now ends up happening is that if all addresses are zero, we return a potentially humongous vector to the caller (in my case it has 201290 elements!), while at the same time reporting a bogus result (the symbol has an address of 0). So in addition to arguably being wrong in certain scenarios, it is a recipe for out of memory situations.

That does not have to be fixed in this very change, but I think we should get to it.

Also, can you please also rebase against `master` and then re-enable `ksym_resolver_load_find`? I disabled it as part of https://github.com/libbpf/blazesym/commit/e1489068dc3b6825ac6921de1344036da6bbc84b

File: src/ksym.rs (hunk 8 above)
Comment (danielocfb) [reply src/ksym.rs:108]: So I added the following test:
```rust
    #[test]
    fn foobar() {
        let mut resolver = KSymResolver::new();
        resolver.syms = vec![
          Ksym { addr: 0x123, name: "1".to_string(), c_name: RefCell::new(None), },
          Ksym { addr: 0x123, name: "1.5".to_string(), c_name: RefCell::new(None), },
          Ksym { addr: 0x1234, name: "2".to_string(), c_name: RefCell::new(None), },
          Ksym { addr: 0x12345, name: "3".to_string(), c_name: RefCell::new(None), },
        ];
        resolver.syms.sort_by(|a, b| a.addr.cmp(&b.addr));

        assert_eq!(resolver.find_addresses_ksym(1), Vec::<&Ksym>::new());
        assert_eq!(resolver.find_addresses_ksym(0x123), vec![
          &Ksym { addr: 0x123, name: "1.5".to_string(), c_name: RefCell::new(None), },
          &Ksym { addr: 0x123, name: "1".to_string(), c_name: RefCell::new(None), },
        ]);

        assert_eq!(resolver.find_addresses_ksym(0x124), Vec::<&Ksym>::new()); // fails
        assert_eq!(resolver.find_addresses_ksym(0x1234568), Vec::<&Ksym>::new()); // fails
    }
```

Is it expected that a search for 0x124 returns symbols at 0x123? What about a search for 0x1234568 returning the symbol at 0x12345? It's not clear to me why that would be desired behavior, but I could be missing something. The documentation seems to suggest the following:
```
    /// Find the names and the start addresses of a symbol found for
    /// the given address.
```
Perhaps I am misreading it, but I am not reading anything about the intention to return potentially completely unrelated symbols.

Assuming the binary search bits check out, I don't really understand the `take_while` condition. Wouldn't something like the following be more appropriate (which incidentally makes the above tests pass, which would certainly be more closely resembling my expectations of the API)?
```patch
--- src/ksym.rs
+++ src/ksym.rs
@@ -109,7 +110,7 @@ impl KSymResolver {
         self.syms[0..l]
             .iter()
             .rev()
-            .take_while(|sym| sym.addr == self.syms[l - 1].addr)
+            .take_while(|sym| sym.addr >= addr)
             .collect()
     }
 }
```

Irrespective of the answer to this last question, I urge you to add a proper set of unit tests with well known inputs so that we can at least have some faith in this code.

Comment (ThinkerYzu1) [reply src/ksym.rs:108]: I am not sure how you test it.  Since I don't implement or derive `Debug` trait, I rewrote the test case you have to test.

```
     #[test]      
    fn foobar() {
        let mut resolver = KSymResolver::new();
        resolver.syms = vec![                                                        
            Ksym { addr: 0x123, name: "1".to_string(), c_name: RefCell::new(None), },
            Ksym { addr: 0x123, name: "1.5".to_string(), c_name: RefCell::new(None), },
            Ksym { addr: 0x1234, name: "2".to_string(), c_name: RefCell::new(None), },
            Ksym { addr: 0x12345, name: "3".to_string(), c_name: RefCell::new(None), },
        ];                                                                                                                              
        //resolver.syms.sort_by(|a, b| a.addr.cmp(&b.addr));                                                                                                                                                                                                                     
                                                            
        assert!(resolver.find_addresses_ksym(1).is_empty());
        let syms = resolver.find_addresses_ksym(0x123);
        assert_eq!(syms.len(), 2);
        assert_eq!(syms[0].addr, 0x123);
        assert_eq!(syms[0].name, "1.5");
        assert_eq!(syms[1].addr, 0x123);
        assert_eq!(syms[1].name, "1");
                                                       
        let syms = resolver.find_addresses_ksym(0x124);
        assert_eq!(syms.len(), 2);
        assert_eq!(syms[0].addr, 0x123);
        assert_eq!(syms[0].name, "1.5");
        assert_eq!(syms[1].addr, 0x123);
        assert_eq!(syms[1].name, "1");                     
        let syms = resolver.find_addresses_ksym(0x1234568);
        assert_eq!(syms.len(), 1);        
        assert_eq!(syms[0].addr, 0x12345);
        assert_eq!(syms[0].name, "3");
    }
```
And,  it pass.

I have mentioned the format of kallsyms in the earlier discussions.  The kallsyms is a list of (start address, symbol name) pairs.  They (pairs) don't include the end addresses or the sizes of symbols or objects.  So, we are finding two symbols that the given address (`addr`) is falling in between their start addresses and returning the earlier one.

For example, `0x123` & `0x124` in your case is falling in between `resolver.syms[1]` and `resolver.syms[2]`, so `resolver.syms[1]` is returned.  However, to handle multiple symbols at the same address, it returns `resolvers.syms[0]` as well.

For the case `0x12345678`, it is bigger than the last symbol.  So, it deems the last symbol covers this address.

Using `sym.addr >= addr` as the `take_while` condition will return all symbols in `self.syms[0..l]`.

For all `i`, `self.syms[i].addr >= addr` will be true where
 - `i < l` ,
 - `self.syms[l - 1].addr <= addr < self.syms[l].addr`, and
 - `self.syms` is sorted.

**Comment by ThinkerYzu1** [comment]
> The test may be passing now, but it's passing despite arguably doing the wrong thing. Because as I said earlier and laid out in https://github.com/libbpf/blazesym/issues/15, we are still treating 0 addresses wrongly. So what now ends up happening is that if all addresses are zero, we return a potentially humongous vector to the caller (in my case it has 201290 elements!),

For external data, it potentially causes out-of-memory even when we are loading data.  However, we can mitigate the symptom.  What do you think to add a limit to the number of returned symbols?

File: src/ksym.rs [hunk 9]
```diff
@@ -91,29 +91,26 @@ impl KSymResolver {
         }
     }
 
-    pub fn find_addresses_ksym(&self, addr: u64) -> Option<&Ksym> {
+    pub fn find_addresses_ksym(&self, addr: u64) -> Vec<&Ksym> {
```
[the remark this answers is not in the document]
Comment (ThinkerYzu1) [reply src/ksym.rs]: fixed

File: src/ksym.rs (hunk 7 above)
Comment (ThinkerYzu1) [reply src/ksym.rs]: fixed

File: src/ksym.rs (hunk 8 above)
Comment (danielocfb) [reply src/ksym.rs:108]: Gotcha. That's kind of what I later came up with as an explanation as well. Can we *please* document that behavior properly, though? I.e., at the very least we should explain that for the very method in question. But I think the `SymResolver` trait should also mention that the results are reported on a best effort basis and some resolvers may return symbols that are not guaranteed to match an address. Ideally we'd encode that better in the type system, but let's do one step at a time.

Comment (danielocfb) [reply src/ksym.rs:108]: Overall the implementation is much cleaner and easier to understand than in the first draft, though. That's great!

**Comment by ThinkerYzu1** [comment]
> Yeah, I'd suggest we should treat 0 as an invalid address. 

Fixed!

File: src/ksym.rs (hunk 8 above)
Comment (ThinkerYzu1) [reply src/ksym.rs:112]: I added the test case to the code.
It tests the cases of addresses that are
 - < first address,
 - == the first address,
 - in between the first and second addresses,
 - == an address in the middle,
 - in between two addresses,
 - == the last address, and
 - &gt;  than the last address.

**Comment by ThinkerYzu1** [comment]
> There is also the possibility of returning iterators as I suggested for a more or less internal API. I think we should consider doing that for the public facing parts 

The majority of our cases return tiny pieces of data.  Iterators don't gain any.  It could even hurt performance in terms of memory locality and branches.  For example, most of the time we return only one or two symbols.  It is very rare to return tens or hundreds of symbols at one time.

File: src/ksym.rs [hunk 10]
```diff
@@ -52,6 +52,10 @@ impl KSymResolver {
             }
             let (addr, _symbol, func) = (tokens[0], tokens[1], tokens[2]);
             if let Ok(addr) = u64::from_str_radix(addr, 16) {
+                if addr == 0 {
+                    line.truncate(0);
```
Comment (danielocfb) [inline src/ksym.rs:56]: I feel like [`String::clear()`](https://doc.rust-lang.org/std/string/struct.String.html#method.clear) conveys intent more succinctly. Consider switching?

**Comment by ThinkerYzu1** [comment]
https://github.com/libbpf/blazesym/actions/runs/3632156556/jobs/6127677966#step:4:119

We should disable this test case for CI since we don't have kallsyms or all addresses are 0.
Is there any way to disable the test case only for CI? A feature?

File: src/ksym.rs [hunk 11]
```diff
@@ -91,29 +95,25 @@ impl KSymResolver {
...
-                r = v;
+                r = m;
             } else {
-                l = v;
+                l = m + 1;
             }
         }
-
-        Some(&self.syms[l])
+        // `self.syms[l - 1].addr <= addr < self.syms[l].addr`.
```
Comment (danielocfb) [inline src/ksym.rs:112]: Nit: Consider removing this line? When I converted it into an assertion said assertion failed with an out of bounds access, so it does not seem kosher as-is. It also raises the question why it's not an assertion and what it really is conveying, given that it's *seemingly* valid Rust code but not actually. I feel like if it is an invariant we could express it in prose or use `debug_assert`.

File: src/ksym.rs (hunk 10 above)
Comment (danielocfb) [reply src/ksym.rs:56]: Please consider adding a unit test triggering this case (without using `/proc/kallsyms`). Can be a follow up change.

**Comment by danielocfb** [comment]
> https://github.com/libbpf/blazesym/actions/runs/3632156556/jobs/6127677966#step:4:119
> 
> We should disable this test case for CI since we don't have kallsyms or all addresses are 0. Is there any way to disable the test case only for CI? A feature?

The easiest is probably to just check if syms is empty and return early. We could introduce a feature or something like that for CI, but this isn't *actually* all that specific to CI.

File: src/ksym.rs (hunk 11 above)
Comment (ThinkerYzu1) [reply src/ksym.rs:112]: Translated to debug_assert().

**Comment by ThinkerYzu1** [comment]
> The easiest is probably to just check if syms is empty and return early. We could introduce a feature or something like that for CI, but this isn't actually all that specific to CI.

I would like to see the error and check my environment to make it successful when I test it locally, instead of failing silently.
In that sense, I would prefer to add a new feature.

**Comment by danielocfb** [comment]
> > The easiest is probably to just check if syms is empty and return early. We could introduce a feature or something like that for CI, but this isn't actually all that specific to CI.
> 
> I would like to see the error and check my environment to make it successful when I test it locally, instead of failing silently. In that sense, I would prefer to add a new feature.

How are you going to fix it? If anybody runs the test suite it should pass, otherwise the test suite is broken and we will end up with bug reports that we can't do anything about. You are free to add a "don't-skip-but-report-an-error" feature for this very purpose, but it should default to off.

File: src/ksym.rs [hunk 12]
```diff
@@ -91,29 +95,25 @@ impl KSymResolver {
...
 
-            if sym.addr == addr {
-                return Some(sym);
-            }
             if addr < sym.addr {
-                r = v;
+                r = m;
             } else {
-                l = v;
+                l = m + 1;
```
Comment (anakryiko) [inline src/ksym.rs:109]: I think this is wrong. `r = m` means that r is supposed to be the correct index at all times (except for cases when there is no element that satisfies condition, which is when all syms have sym.addr > addr; but this corner case is handled properly in the loop after this binary search (at least in my example). But `addr < sym.addr` is not the condition we are looking for, we want to find rightmost element that satisfies addr >= sym.addr.

Please take a look at my examples from the discussion before. I think they are more correct (though there could still be bugs I missed, but let's review that).

**Review by anakryiko — CHANGES_REQUESTED** [review]

File: src/ksym.rs [hunk 13]
```diff
@@ -91,29 +95,25 @@ impl KSymResolver {
...
             }
         }
-
-        Some(&self.syms[l])
+        // `self.syms[l - 1].addr <= addr < self.syms[l].addr`.
+        self.syms[0..l]
+            .iter()
+            .rev()
+            .take_while(move |sym| sym.addr == self.syms[l - 1].addr)
```
[the remark this answers is not in the document]
Comment (ThinkerYzu1) [reply src/ksym.rs:116]: The test case is already added.
if `l` is 0, the closure will never be executed since self.syms[0..0] is empty.  If `l` is `1`, it will compare to itself. I means `sym` is identical to `self.syms[l - 1]`.

File: src/ksym.rs [hunk 14]
```diff
@@ -91,29 +95,25 @@ impl KSymResolver {
         }
     }
 
-    pub fn find_address_ksym(&self, addr: u64) -> Option<&Ksym> {
+    pub fn find_addresses_ksym(&self, addr: u64) -> impl Iterator<Item = &Ksym> {
         let mut l = 0;
         let mut r = self.syms.len();
```
[the remark this answers is not in the document]
Comment (ThinkerYzu1) [reply src/ksym.rs:100]: The range here is `[l..r)`.  Doesn't include `r`.

**Comment by ThinkerYzu1** [comment]
int n = self.syms.len();
    int l = 0, r = n - 1;
    while (l < r) {
      int m = l + (r - l) / 2;
      if (self.syms[m].addr < addr)
        l = m + 1;
      else
        r = m;
    }

Let's look at the previous example.

The first issue is if the address is falling in between two symbols, it will stop at the position where `syms[l - 1].addr < addr < syms[l].addr`.  Then, it will return symbols on the left side.  However, if the address is matched exactly, it stop at the position where `syms[l - 1] < addr == syms[l].addr`, it should return symbols on the right side.

The other issue is if the address we are searching for is bigger than the last address, the it will stop at the position where `l == self.syms.len() - 1` since `r = self.syms.len() - 1` and `l` is never bigger than `r`.  At this point, `addr >= self.syms[l].addr`, and potentially `self.syms[l].addr == self.syms[l - 1]` as well since multiple symbols can be at the same address.  That means it should return both symbols on the left side (`l-1, l-2, ....`)and the one on the spot (`l`).  It is inconsistent with the cases mentioned in the first issue.

What if we change it to `if (self.syms[m].addr < addr)`?  It will fix the first issue.  However, the second issue is still there.

**Review by anakryiko — APPROVED** [review]

**Comment by ThinkerYzu1** [comment]
#include <stdio.h>
    
    int search(const int *data, int n, int v) {
            int l = 0, r = n - 1;
            while (l < r) {
                    printf("(l=%d, r=%d) ", l, r);
                    int m = l + (r - l) / 2;
                    if (data[m] < v)
                            l = m + 1;
                    else
                            r = m;
            }
            printf("(l=%d, r=%d) ", l, r);
            return l;
    }
    
    int main() {
            int data[] = { 1, 1, 1, 5, 5 };
            int v;
            v = 2;
            printf("%d %d\n", v, search(data, 5, v));
            v = 5;
            printf("%d %d\n", v, search(data, 5, v));
            v = 7;
            printf("%d %d\n", v, search(data, 5, v));
    }

I try to rewrite the algorithm in C.  And run the test cases mentioned above.

**Comment by anakryiko** [comment]
nice, thank you for specific examples, this way we can finally come to some conclusion. I think you are using a wrong version of my code in your analysis. I took the very first example I wrote in that long discussion, I change literally nothing about the binary search while loop, and I adjusted for loop to go backwards (as we discussed in that thread). I also added l, r printing and what are the final returned results (so output is a bit different from your example). Please check if it makes more sense now:

```
$ cat temp.c
#include <stdio.h>

void doit(int *a, int n, int addr)
{
        int i, l = 0, r = n - 1;
        while (l < r) {
                printf("(l=%d, r=%d) ", l, r);
                int m = l + (r - l + 1) / 2;
                if (a[m] <= addr)
                        l = m;
                else
                        r = m - 1;
        }
        printf("(l=%d, r=%d) ", l, r);
        for (i = l; i >= 0 && a[i] == a[l]; i--) {
                printf("%d ", a[i]);
        }
        printf("\n");
}

int main() {
        int arr[] = {1, 1, 1, 5, 5};

        doit(arr, 5, 2);
        doit(arr, 5, 5);
        doit(arr, 5, 7);

        return 0;
}
$ cc -g temp.c && ./a.out
(l=0, r=4) (l=2, r=4) (l=2, r=2) 1 1 1
(l=0, r=4) (l=2, r=4) (l=3, r=4) (l=4, r=4) 5 5
(l=0, r=4) (l=2, r=4) (l=3, r=4) (l=4, r=4) 5 5
```

I think all the results are correct.

**Comment by anakryiko** [comment]
I don't know why you used my second example if we later on agreed that the first example had the right semantics (except for for loop that produces results). Have you seen my reply in https://github.com/libbpf/blazesym/pull/23#discussion_r1037498294?

**Comment by ThinkerYzu1** [comment]
I think we are on the page now. I initialize `r` to `self.syms.len()` while you initialize it to `self.syms.len() - 1` but add `1` back when computing `m`.

**Comment by anakryiko** [comment]
So I still made a mistake in that for() loop, there should be another condition:

```
for (i = l; i < n && a[i] <= addr && a[i] == a[l]; i--)
```

But I also went ahead and wrote a simple exhaustive search over all partial ordering of up to 10 elements. This goes and generates arrays like [1,1,1], [1,1,2], [1,2,2], [1,2,3] and so on. And then for each such array it searched each possible value between 0 and N+1, which I think should be able to generate all the corner cases. Please see below. It would be great to write something like this for ksym as a unit test case. I don't think we need to go all the way to 10 element arrays, testing up to 5-6 elements should be fine, I think.

But here it is for reference (just in case I made yet another mistake somewhere, maybe you'll catch it):

```
$ cat temp.c
#include <stdio.h>

void doit(int *a, int n, int addr, int *res_l, int *res_r)
{
        int i, l = 0, r = n - 1;
        while (l < r) {
                int m = l + (r - l + 1) / 2;
                if (a[m] <= addr)
                        l = m;
                else
                        r = m - 1;
        }
        *res_l = -1;
        *res_r = -1;
        for (i = l; i < n && a[i] <= addr && a[i] == a[l]; i--) {
                if (*res_r == -1)
                        *res_r = i;
                *res_l = i;
        }
}

void doit_simple(int *a, int n, int addr, int *res_l, int *res_r)
{
        int i;

        *res_l = -1;
        *res_r = -1;
        for (i = n - 1; i >= 0; i--) {
                if (a[i] <= addr) {
                        *res_r = i;
                        *res_l = i;
                        break;
                }
        }
        for (i = i - 1; i >= 0 && a[i] == a[*res_r]; i--) {
                *res_l = i;
        }
}

static int a[10];
static int cnt;

void recur(int N, int p)
{
        int i, j;

        if (p == N) {
                int res_l, res_r;
                int ctr_l, ctr_r;
                /* we generated partial ordering
                 * now let's validate finding all possible "addresses"
                 * in the range of [0, N+1]
                 */

                for (i = 0; i <= N + 1; i++) {
                        doit(a, N, i, &res_l, &res_r);
                        doit_simple(a, N, i, &ctr_l, &ctr_r);
                        cnt++;

                        if (res_l != ctr_l || res_r != ctr_r) {
                                printf("VIOLATION!!! N=%d, addr=%d, res_l=%d, res_r=%d, ctr_l=%d, ctr_r=%d arr=[",
                                        N, i, res_l, res_r, ctr_l, ctr_r);
                                for (j = 0; j < N; j++)
                                        printf("%d,", a[j]);
                                printf("]\n");
                        }
                }

                return;
        }

        /* each element of an array is in [1, N] range */
        for (i = (p == 0 ? 1 : a[p - 1]); i <= N; i++) {
                a[p] = i;
                recur(N, p + 1);
        }
}

int main() {
        int arr[] = {1, 1, 1, 5, 5};
        int N;

        for (N = 0; N <= 10; N++)
                recur(N, 0);
        printf("VALIDATED %d cases\n", cnt);
        return 0;
}
$ cc -g temp.c && ./a.out
VALIDATED 1460595 cases
```

It took 0.07s in debug mode, so I think it can be easily done as a unit test that runs every single time in CI.

**Comment by anakryiko** [comment]
I translated your implementation in Rust to equivalent C (I think :) please double-check) like below. It also passes all tests. So please consider implementing this exhaustive test as a proper unit test.

```
void doit(int *a, int n, int addr, int *res_l, int *res_r)
{
        int i, l = 0, r = n;
        while (l < r) {
                int m = (l + r) / 2;
                if (addr < a[m])
                        r = m;
                else
                        l = m + 1;
        }
        *res_l = -1;
        *res_r = -1;
        for (i = l - 1; i >= 0 && i < n && a[i] == a[l - 1]; i--) {
                if (*res_r == -1)
                        *res_r = i;
                *res_l = i;
        }
}
```

**Comment by ThinkerYzu1** [comment]
Added an exhaustive test case.  Debug build is slow.  It takes 2s at local.  The release build is reasonably fast.  The number of test cases is still small now.  We can afford it yet.

**Comment by ThinkerYzu1** [comment]
Thank you for the review.
